crates/rbuilder/src/live_builder/building/unfinished_block_building_sink_muxer.rs

```diff
@@ -0,0 +1,243 @@
+use std::sync::{Arc, Mutex};
+
+use ahash::HashMap;
+use alloy_primitives::TxHash;
+
+use crate::building::builders::{
+    block_building_helper::BiddableUnfinishedBlock, UnfinishedBlockBuildingSink,
+};
+
+/// Multiplexes the input from several building algos by forwarding best block among the last blocks
+/// received from each algo. Algos are identified by block().builder_name()
+pub struct UnfinishedBlockBuildingSinkMuxer {
+    /// needed state to check and update the best block.
+    best_block_state: Mutex<BestBlockState>,
+    /// Here we push the best blocks.
+    sink: Arc<dyn UnfinishedBlockBuildingSink>,
+}
+
+impl UnfinishedBlockBuildingSinkMuxer {
+    pub fn new(sink: Arc<dyn UnfinishedBlockBuildingSink>) -> Self {
+        Self {
+            best_block_state: Mutex::new(BestBlockState::new()),
+            sink,
+        }
+    }
+}
+
+impl UnfinishedBlockBuildingSink for UnfinishedBlockBuildingSinkMuxer {
+    fn new_block(&self, block: BiddableUnfinishedBlock) {
+        if let Some(best_biddable_block) = self
+            .best_block_state
+            .lock()
+            .unwrap()
+            .update_best_block(block)
+        {
+            self.sink.new_block(best_biddable_block);
+        }
+    }
+
+    fn can_use_suggested_fee_recipient_as_coinbase(&self) -> bool {
+        self.sink.can_use_suggested_fee_recipient_as_coinbase()
+    }
+}
+
+impl std::fmt::Debug for UnfinishedBlockBuildingSinkMuxer {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        f.debug_struct("UnfinishedBlockBuildingSinkMuxer").finish()
+    }
+}
+
+/// Computes the best block among all the algos.
+struct BestBlockState {
+    /// For each algo name we store the last block so on each update we can look for the best block among all algos.
+    /// We should replace String for "real" type identifying the algo.
+    algorithm_to_last_block: HashMap<String, BiddableUnfinishedBlock>,
+    /// Works as an id for the last block so we don't send the same twice since the same block could be generated by different algos.
+    last_sent_block_txs: Option<Vec<TxHash>>,
+}
+
+impl BestBlockState {
+    fn new() -> Self {
+        Self {
+            algorithm_to_last_block: Default::default(),
+            last_sent_block_txs: None,
+        }
+    }
+
+    /// Given a new BiddableUnfinishedBlock updates the best per algo and return the best one if
+    /// changes from last time.
+    fn update_best_block(
+        &mut self,
+        biddable_block: BiddableUnfinishedBlock,
+    ) -> Option<BiddableUnfinishedBlock> {
+        self.algorithm_to_last_block.insert(
```

Comment:
nit: allocating string every time here is not necessary as its going to be filled and replaced